File: fibril/webdemo/p5-fibril-viz/fibril-algorithm.js
// Fibril Algorithm - Node.js Module
// Bayesian MIDI Note Selection Algorithm

const MIDI_SIZE = 128;
const NUM_ITERATIONS = 3;
const MAJOR_SCALE_INTERVALS = [0, 2, 4, 5, 7, 9, 11];

/**
 * Seeded random number generator for reproducibility
 */
class SeededRandom {
    constructor(seed) {
        this.seed = seed;
    }

    next() {
        this.seed = (this.seed * 9301 + 49297) % 233280;
        return this.seed / 233280;
    }
}

/**
 * Generate major key note classes from root note
 * @param {number} rootNote - MIDI number of the root (0-127)
 * @returns {number[]} Array of note classes (0-11) in the major key
 */
function generateMajorKey(rootNote) {
    return MAJOR_SCALE_INTERVALS.map(interval => (rootNote + interval) % 12);
}

/**
 * Check if a MIDI note is in the specified key
 * @param {number} midiNote - MIDI note number (0-127)
 * @param {number[]} majorKey - Array of note classes in key
 * @returns {boolean}
 */
function isInKey(midiNote, majorKey) {
    const noteClass = midiNote % 12;
    return majorKey.includes(noteClass);
}

/**
 * Create a binary mask for in-key notes
 * @param {number} rootNote - MIDI root note
 * @returns {number[]} Array of 0s and 1s
 */
function createInKeyMask(rootNote) {
    const majorKey = generateMajorKey(rootNote);
    const mask = new Array(MIDI_SIZE);
    for (let i = 0; i < MIDI_SIZE; i++) {
        mask[i] = isInKey(i, majorKey) ? 1 : 0;
    }
    return mask;
}

/**
 * COMPUTE function: Calculate harmonic relationship score
 * @param {number} bIndex - POSTERIOR index being calculated
 * @param {number} aIndex - UIN_MAP index being compared
 * @param {number[]} voicingMap - Binary array of voiced notes
 * @param {number[]} inKeyMask - Binary array of in-key notes
 * @returns {number} Harmonic score
 */
function compute(bIndex, aIndex, voicingMap, inKeyMask) {
    // Rule 1: Key filtering
    if (inKeyMask[bIndex] === 0) {
        return 0.0;
    }

    let score = 0.0;
    const intervalWithA = bIndex - aIndex;

    // Rule 2: Perfect fourths (interval = ±5)
    if (Math.abs(intervalWithA) === 5) {
        score += 0.3;
    }

    // Rule 3: Perfect fifths (interval = ±7) with A
    if (Math.abs(intervalWithA) === 7) {
        score += 0.25;
    }

    // Check against all voiced notes
    for (let voicedNote = 0; voicedNote < MIDI_SIZE; voicedNote++) {
        if (voicingMap[voicedNote] === 1) {
            const intervalWithVoiced = bIndex - voicedNote;

            // Rule 2: Perfect fourths with voiced notes
            if (Math.abs(intervalWithVoiced) === 5) {
                score += 0.3;
            }

            // Rule 3: Perfect fifths with voiced notes
            if (Math.abs(intervalWithVoiced) === 7) {
                score += 0.25;
            }

            // Rule 4: Voice leading (interval = ±1 or ±2)
            if (Math.abs(intervalWithVoiced) === 1 || Math.abs(intervalWithVoiced) === 2) {
                score += 0.2;
            }
        }
    }

    return score;
}

/**
 * Perform one iteration of Bayesian update
 * @param {number[]} posterior - Current probability distribution
 * @param {number[]} uinMap - Prior distribution
 * @param {number[]} voicingMap - Voiced notes
 * @param {number[]} inKeyMask - In-key mask
 * @returns {number[]} Updated posterior distribution
 */
function bayesianUpdate(posterior, uinMap, voicingMap, inKeyMask) {
    const newPosterior = new Array(MIDI_SIZE);

    // For each element j in POSTERIOR
    for (let j = 0; j < MIDI_SIZE; j++) {
        let sum = 0.0;

        // Sum over all elements i in UIN_MAP
        for (let i = 0; i < MIDI_SIZE; i++) {
            const pA = uinMap[i]; // Prior probability
            const pBGivenA = compute(j, i, voicingMap, inKeyMask); // Likelihood
            const pB = posterior[j]; // Normalizing constant

            // Avoid division by zero
            if (pB > 0) {
                sum += (pA * pBGivenA) / pB;
            }
        }

        newPosterior[j] = sum;
    }

    // Normalize to sum to 1
    let total = 0.0;
    for (let j = 0; j < MIDI_SIZE; j++) {
        total += newPosterior[j];
    }

    if (total > 0) {
        for (let j = 0; j < MIDI_SIZE; j++) {
            newPosterior[j] /= total;
        }
    }

    return newPosterior;
}

/**
 * Sample from probability distribution using seeded random
 * @param {number[]} posterior - Probability distribution
 * @param {SeededRandom} rng - Random number generator
 * @returns {number} Selected MIDI note index
 */
function sampleFromDistribution(posterior, rng) {
    const randValue = rng.next();
    let cumulative = 0.0;

    for (let i = 0; i < MIDI_SIZE; i++) {
        cumulative += posterior[i];
        if (randValue <= cumulative) {
            return i;
        }
    }

    return MIDI_SIZE - 1; // Fallback
}

/**
 * Allocate voices based on probability distribution
 * @param {number[]} posterior - Probability distribution
 * @param {number} numVoices - Number of voices to allocate
 * @param {number} seed - Random seed
 * @returns {number[]} Voicing map (binary array)
 */
function allocateVoices(posterior, numVoices, seed) {
    const voicingMap = new Array(MIDI_SIZE).fill(0);
    const rng = new SeededRandom(seed);

    for (let v = 0; v < numVoices; v++) {
        const selectedNote = sampleFromDistribution(posterior, rng);
        voicingMap[selectedNote] = 1;
    }

    return voicingMap;
}

/**
 * Main Fibril algorithm
 * @param {Object} params - Algorithm parameters
 * @param {number} params.numVoices - Number of voices to allocate
 * @param {number} params.seed - Random seed
 * @param {number} params.rootNote - Root note MIDI number
 * @returns {Object[]} Array of states at each stage
 */
function runFibrilAlgorithm({ numVoices, seed, rootNote }) {
    const inKeyMask = createInKeyMask(rootNote);

    // Initialize POSTERIOR with uniform distribution
    const initialProb = 1.0 / MIDI_SIZE;
    let posterior = new Array(MIDI_SIZE).fill(initialProb);
    let voicingMap = new Array(MIDI_SIZE).fill(0);

    // Store initial state
    const states = [
        { stage: 'Initial', posterior: [...posterior], voicingMap: [...voicingMap] }
    ];

    // Apply in-key mask to initial state
    for (let i = 0; i < MIDI_SIZE; i++) {
        posterior[i] *= inKeyMask[i];
    }

    // Normalize after masking
    let total = 0.0;
    for (let i = 0; i < MIDI_SIZE; i++) {
        total += posterior[i];
    }
    if (total > 0) {
        for (let i = 0; i < MIDI_SIZE; i++) {
            posterior[i] /= total;
        }
    }

    // Run comparison loop (3 iterations)
    for (let iteration = 0; iteration < NUM_ITERATIONS; iteration++) {
        // Copy POSTERIOR to UIN_MAP
        const uinMap = [...posterior];

        // Perform Bayesian update
        posterior = bayesianUpdate(posterior, uinMap, voicingMap, inKeyMask);

        // Store state after this iteration
        states.push({
            stage: `Iteration ${iteration + 1}`,
            posterior: [...posterior],
            voicingMap: [...voicingMap]
        });
    }

    // Allocate voices after all iterations
    voicingMap = allocateVoices(posterior, numVoices, seed);

    // Update final state with allocated voices
    states[states.length - 1].voicingMap = voicingMap;

    return states;
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        runFibrilAlgorithm,
        MIDI_SIZE,
        NUM_ITERATIONS
    };
}

File: fibril/webdemo/fibril_algorithm.lua
-- Fibril Algorithm: Bayesian MIDI Note Selection
-- Lua Pseudocode Implementation

-- Constants
local MIDI_SIZE = 128
local NUM_ITERATIONS = 3
local INITIAL_PROBABILITY = 1.0 / MIDI_SIZE

-- User configuration
local config = {
    key = "C",  -- Musical key
    num_voices = 4  -- Number of voices to allocate
}

-- C Major scale MIDI notes (in one octave, pattern repeats)
local C_MAJOR_SCALE = {0, 2, 4, 5, 7, 9, 11}  -- C, D, E, F, G, A, B

-- Check if a MIDI note is in the specified key
function is_in_key(midi_note, key)
    -- For simplicity, assuming key is "C"
    local note_class = midi_note % 12
    for _, scale_note in ipairs(C_MAJOR_SCALE) do
        if note_class == scale_note then
            return true
        end
    end
    return false
end

-- Calculate harmonic boost based on interval
function get_interval_boost(interval, boost_type)
    local abs_interval = math.abs(interval)

    if boost_type == "fourth" and abs_interval == 5 then
        return 0.3
    elseif boost_type == "fifth" and abs_interval == 7 then
        return 0.25
    elseif boost_type == "voice_leading" and (abs_interval == 1 or abs_interval == 2) then
        return 0.2
    end

    return 0.0
end

-- COMPUTE function: Calculate harmonic relationship score
function compute(b_index, a_index, voicing_map, key)
    -- Rule 1: Key filtering
    if not is_in_key(b_index, key) then
        return 0.0
    end

    local score = 0.0

    -- Rule 2: Perfect fourths
    -- Check against A_index
    local interval_with_a = b_index - a_index
    score = score + get_interval_boost(interval_with_a, "fourth")

    -- Check against all voiced notes in VOICING_MAP
    for voiced_note = 0, MIDI_SIZE - 1 do
        if voicing_map[voiced_note] == 1 then
            local interval_with_voiced = b_index - voiced_note
            score = score + get_interval_boost(interval_with_voiced, "fourth")
        end
    end

    -- Rule 3: Perfect fifths
    -- Check against A_index
    score = score + get_interval_boost(interval_with_a, "fifth")

    -- Check against all voiced notes
    for voiced_note = 0, MIDI_SIZE - 1 do
        if voicing_map[voiced_note] == 1 then
            local interval_with_voiced = b_index - voiced_note
            score = score + get_interval_boost(interval_with_voiced, "fifth")
        end
    end

    -- Rule 4: Voice leading
    for voiced_note = 0, MIDI_SIZE - 1 do
        if voicing_map[voiced_note] == 1 then
            local interval_with_voiced = b_index - voiced_note
            score = score + get_interval_boost(interval_with_voiced, "voice_leading")
        end
    end

    return score
end

-- Sample from probability distribution using random seed
function sample_from_distribution(posterior, random_seed)
    math.randomseed(random_seed)
    local rand_value = math.random()

    local cumulative = 0.0
    for i = 0, MIDI_SIZE - 1 do
        cumulative = cumulative + posterior[i]
        if rand_value <= cumulative then
            return i
        end
    end

    return MIDI_SIZE - 1  -- Fallback
end

-- Initialize arrays
function initialize()
    local posterior = {}
    local uin_map = {}
    local voicing_map = {}

    for i = 0, MIDI_SIZE - 1 do
        posterior[i] = INITIAL_PROBABILITY
        uin_map[i] = INITIAL_PROBABILITY
        voicing_map[i] = 0
    end

    return posterior, uin_map, voicing_map
end

-- Perform one comparison iteration (Bayesian update)
function comparison_iteration(posterior, uin_map, voicing_map, key)
    local new_posterior = {}

    -- For each element j in POSTERIOR
    for j = 0, MIDI_SIZE - 1 do
        local sum = 0.0

        -- Sum over all elements i in UIN_MAP
        for i = 0, MIDI_SIZE - 1 do
            -- Bayes's theorem: P(A|B) = P(A) * P(B|A) / P(B)
            local p_a = uin_map[i]  -- Prior probability
            local p_b_given_a = compute(j, i, voicing_map, key)  -- Likelihood
            local p_b = posterior[j]  -- Normalizing constant

            -- Avoid division by zero
            if p_b > 0 then
                sum = sum + (p_a * p_b_given_a / p_b)
            end
        end

        new_posterior[j] = sum
    end

    -- Normalize the posterior to ensure it sums to 1
    local total = 0.0
    for j = 0, MIDI_SIZE - 1 do
        total = total + new_posterior[j]
    end

    if total > 0 then
        for j = 0, MIDI_SIZE - 1 do
            new_posterior[j] = new_posterior[j] / total
        end
    end

    return new_posterior
end

-- Perform three iterations of the comparison loop
function run_comparison_loop(posterior, voicing_map, key)
    for iteration = 1, NUM_ITERATIONS do
        -- Copy POSTERIOR to UIN_MAP at start of iteration
        local uin_map = {}
        for i = 0, MIDI_SIZE - 1 do
            uin_map[i] = posterior[i]
        end

        -- Perform Bayesian update
        posterior = comparison_iteration(posterior, uin_map, voicing_map, key)

        print("Iteration " .. iteration .. " complete")
    end

    return posterior
end

-- Allocate voices based on probability distribution
function allocate_voices(posterior, voicing_map, num_voices)
    -- Count currently allocated voices
    local current_voices = 0
    local allocated_indices = {}

    for i = 0, MIDI_SIZE - 1 do
        if voicing_map[i] == 1 then
            current_voices = current_voices + 1
            table.insert(allocated_indices, i)
        end
    end

    print("Current voices: " .. current_voices .. ", Target: " .. num_voices)

    -- Case 1: Need to allocate more voices
    if num_voices > current_voices then
        -- Allocate new voices
        for v = current_voices + 1, num_voices do
            local random_seed = os.time() + v
            local selected_index = sample_from_distribution(posterior, random_seed)
            voicing_map[selected_index] = 1
            print("Allocated voice " .. v .. " at MIDI " .. selected_index)
        end

        -- Optionally: free and reroll old voices one at a time
        -- (Implementation detail depends on specific requirements)

    -- Case 2: Need to free voices
    elseif num_voices < current_voices then
        -- Free voices one at a time until quota is reached
        for i = 1, current_voices - num_voices do
            if allocated_indices[i] then
                voicing_map[allocated_indices[i]] = 0
                print("Freed voice at MIDI " .. allocated_indices[i])
            end
        end
    end

    return voicing_map
end

-- Main algorithm
function fibril_algorithm()
    print("=== Fibril Algorithm ===")
    print("Initializing...")

    -- Initialize data structures
    local posterior, uin_map, voicing_map = initialize()

    -- Run the comparison loop (3 iterations)
    print("\nRunning comparison loop...")
    posterior = run_comparison_loop(posterior, voicing_map, config.key)

    -- Allocate voices
    print("\nAllocating voices...")
    voicing_map = allocate_voices(posterior, voicing_map, config.num_voices)

    -- Display results
    print("\n=== Results ===")
    print("Top 10 most probable notes:")

    -- Sort by probability (simplified - just showing concept)
    for rank = 1, 10 do
        local max_prob = 0.0
        local max_index = 0

        for i = 0, MIDI_SIZE - 1 do
            if posterior[i] > max_prob then
                max_prob = posterior[i]
                max_index = i
            end
        end

        print(rank .. ". MIDI " .. max_index .. ": " .. string.format("%.6f", max_prob))
        posterior[max_index] = 0  -- Remove from next search
    end

    print("\nAllocated voices (VOICING_MAP):")
    for i = 0, MIDI_SIZE - 1 do
        if voicing_map[i] == 1 then
            print("  MIDI " .. i)
        end
    end
end

-- Run the algorithm
fibril_algorithm()
